### DISCUSSION TITLE
Escaping special tokens (e.g. colon) in command name?

### DISCUSSION THREAD

**Discussion by joshuatz** [casey/just#1074] [Q&A] [answered]
I searched through all open and closed issues, as well as the docs, and it surprisingly seems as though this has never come up before (if so, apologies).

Is it possible to craft a command name that contains special characters / tokens that otherwise break `just` since they are parsed as such? It looks like multiple characters will cause errors if included as part of a command name (such as `=`, `&`, etc.), but I am personally most interested in learning if it would be possible to use `:` as part of a command name.

The colon character is frequently used in Node development, and you will often find `package.json` script entries that look like this:

```json
{
  "scripts": {
    "build:prod": "build.js --prod --minify",
    "build:dev": "build.js --dev --source-map --verbose"
  }
}
```

Where running `npm run build:prod` works without issue.

If you try to replicate these names in `just`, you will immediately run into issues:

```
build:prod:
  build.js --prod --minify
```

> ❌ error: Expected '&&', comment, end of file, end of line, identifier, or '(', but found ':'

Is there a way around this? I realize that switching from `build:prod` to `build-prod`, or more dynamic `build $target` or `build TARGET` would be an immediate solution and perhaps even better practice with `just`, but I wanted to put this question out there in case A) there was a solution, and B) in case there are other devs hitting this while trying to replicate `package.json` scripts.

I do see some docs on escaping, but they seem to only apply to the body of commands, not the command name.

**Comment by casey** [comment, marked as answer]
Good question!

Currently, recipe names must be valid `just` identifiers, which are of the form `[a-zA-Z_][a-zA-Z0-9_-]*`, so it isn't possible to use `:` in a recipe name.

Since the `package.json` file is JSON, and script names are Javascript strings, there aren't any restrictions on their contents, whereas `just` recipe names have to be distinguished from other parts of the `just` language, i.e. the lexer needs to be able to tell where a recipe name ends and the `:` after the name begins, so there are more restrictions on recipe name contents.

I would suggest using dashes, i.e., `foo-bar`. It's a pinky reach, but you don't have to press the shift key, so it's kind of a wash.

**Reply by joshuatz** [reply]
Gotcha, that makes sense. And to be fair to `just`, it definitely isn't alone in that rule - seems like a lot of task / command running tools that use config files based on or borrowing from YAML syntax all (understandably, due to the significance of `:` in that context) have that limitation.

> I would suggest using dashes, i.e., `foo-bar`

Yep, agreed. It's not a huge deal for me, but I'm really glad to get an official answer here!

Thanks for all your hard work on this tool and your quick and detailed response here!
